### DISCUSSION TITLE
How do I use discord.py in a non-interactive way (without loop)?

### DISCUSSION THREAD

**Discussion by honzajavorek** [Rapptz/discord.py#6278] [Q&A] [answered]
Hi, my intention is to make a simple, one-time script, which connects to my discord server, downloads information about my members, and finishes. That's it, no loop, no interactivity. Only I decide when I manually run it. I couldn't find any example for that. Even the [background task](https://github.com/Rapptz/discord.py/blob/async/examples/background_task.py) example hangs in the loop. I'm new to all the async/await stuff in Python, so perhaps the answer isn't specific to discord.py, but nevertheless so far I'm lost. Could someone give me some guidance? Thank you!

**Comment by Rapptz** [comment, marked as answer]
Run your code and log off within `on_ready`.

e.g.

```py
class Client(discord.Client):
    async def on_ready(self):
        # my code here...
        await self.close()

client = Client()
client.run('token')
```

I would have recommended webhooks which are more lightweight if your use case was just sending messages but it seems you want to do more and it seems to involve the gateway machinery as well so this is probably the easiest thing to do what you want.

**Reply by honzajavorek** [reply]
Seems to work, thanks!

**Reply by honzajavorek** [reply]
After I grappled a bit with error handling, I wrote it all down today in an article: [How to create a non-interactive Discord bot](https://honzajavorek.cz/blog/how-to-create-non-interactive-discord-bot/) I hope it helps someone else to avoid my mistakes.

**Reply by bandwiches** [reply]
Nice write up!  FWIW this repo was my intro into async/await and now I find myself trying to use it everywhere.

**Reply by honzajavorek** [reply]
Yup, it's my first serious-ish experience with asyncio, too.

**Reply by 15696** [reply]
@honzajavorek Hmm.. Why do you need to wait_until_ready if you're running the code in on_ready?

**Reply by 15696** [reply]
Another typo I noticed.
![image](https://user-images.githubusercontent.com/50557523/107156434-87f68880-693b-11eb-9994-05efc43ec1bc.png)

**Reply by honzajavorek** [reply]
@mudkipdev It's a few weeks between writing the script and writing the article, so TBH I don't know why the waiting is needed and perhaps it's not. I'd have to try and confirm. The `pprint()` is not a typo, that's correct: [pprint()](https://docs.python.org/3/library/pprint.html)
